packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,226 @@
+    ///         considered unsafe (see the _isUnsafeTarget() function).
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _target      Target contract or wallet address.
+    /// @param _message     Message to trigger the target address with.
+    function sendMessage(uint256 _destination, address _target, bytes calldata _message) external payable {
+        if (_destination == block.chainid) revert MessageDestinationSameChain(_destination);
+
+        bytes memory data = abi.encodeCall(
+            L2ToL2CrossDomainMessenger.relayMessage,
+            (_destination, block.chainid, messageNonce(), msg.sender, _target, _message)
+        );
+        emit SentMessage(data);
+        msgNonce++;
+    }
+
+    /// @notice Relays a message that was sent by the other CrossDomainMessenger contract. Can only
+    ///         be executed via cross-chain call from the other messenger OR if the message was
+    ///         already received once and is currently being replayed.
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _source      Chain ID of the source chain.
+    /// @param _nonce       Nonce of the message being relayed.
+    /// @param _sender      Address of the user who sent the message.
+    /// @param _target      Address that the message is targeted at.
+    /// @param _message     Message to send to the target.
+    function relayMessage(
+        uint256 _destination,
+        uint256 _source,
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        bytes memory _message
+    )
+        external
+        payable
+        nonReentrant
+    {
+        if (msg.sender != Predeploys.CROSS_L2_INBOX) revert RelayCallerNotCrossL2Inbox(msg.sender);
+        {
+            address origin = CrossL2Inbox(Predeploys.CROSS_L2_INBOX).origin();
+            if (origin != Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) {
+                revert CrossL2InboxOriginNotL2ToL2CrossDomainMessenger(origin);
+            }
+        }
+        if (_destination != block.chainid) revert MessageDestinationNotRelayChain(_destination);
+        if (_target == Predeploys.CROSS_L2_INBOX) revert MessageTargetCrossL2Inbox();
+        if (_target == Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) {
+            revert MessageTargetL2ToL2CrossDomainMessenger();
+        }
+
+        bytes32 messageHash = keccak256(abi.encode(_destination, _source, _nonce, _sender, _target, _message));
+        if (successfulMessages[messageHash] == true) {
+            revert MessageAlreadyRelayed(messageHash);
+        }
+
+        _storeMessageMetadata(_source, _sender);
+
+        bool success = _callWithAllGas(_target, _message);
+
+        if (success) {
+            successfulMessages[messageHash] = true;
+            emit RelayedMessage(messageHash);
+        } else {
+            emit FailedRelayedMessage(messageHash);
+        }
```

Comment:
Not suggesting that we do this but something to think about. You could imagine if there was a failed messages mapping, you could skip the auth checks above (all of the reverts) and just relay the call again and then delete from the failed messages mapping if it succeeds.

Not sure if this is super useful and seems like a way to introduce a bug

Reply:
I like this idea, but it would introduce significant additional logic to `relayMessage` (with risk of bug as you said). should we still go ahead?

Reply:
we should not do this but just thinking out loud :)